Improved error codes in case of security exception

vamsi-amazon opened the pull request:
### Description
Currently all the security exceptions related to datasource APIs are thrown as internal server errors.
With this change, the error message will respect the status received from upstream core service/plugins.
 
### Issues Resolved
[List any issues this PR will resolve]
 
### Check List
- [ ] New functionality includes testing.
  - [ ] All tests pass, including unit test, integration test and doctest
- [x] New functionality has been documented.
  - [x] New functionality has javadoc added
  - [x] New functionality has user manual doc added
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

Yury-Fridlyand reviewed (commented):
LGFM but CI fails

vamsi-amazon commented:
> LGFM but CI fails

CI Failures are unrelated to the above changes.
Integ tests issue: https://github.com/opensearch-project/sql/issues/1754

Yury-Fridlyand commented:
CI failed due to lack of jacoco coverage!
```
* What went wrong:
Execution failed for task ':opensearch:test'.
> There were failing tests. See the report at: file:///home/runner/work/sql/sql/opensearch/build/reports/tests/test/index.html
```

vamsi-amazon commented:
@Yury-Fridlyand  UT failures are due to ML-Commons. I have verified by running without my commit before merging.
These lines are not configured under Jacoco and it won't cause build failure. 

I have added unit tests but it requires opensearch.test framework and it is disrupting the mockito runtime libraries and causing issues in all the existing test cases. Covering with integ tests is not possible as the error scenario requires security plugin. I have created an issue to https://github.com/opensearch-project/sql/issues/1754 handle this.

Added new issue for jacoco coverage of rest handlers: https://github.com/opensearch-project/sql/issues/1756

Yury-Fridlyand commented:
If your PR requires other changes to have CI passed, do it in scope of this PR or in another PR BEFORE merging this one.

Yury-Fridlyand commented:
Appealing to your experience @dblock, can we leave this as is and hope for a coming fix or this PR should be reverted?

vamsi-amazon commented:
Would like to hear his thoughts. I will try to fix ML commons error. If I could not, i will raise a revert PR by EOD.

vamsi-amazon commented:
The error is due to https://github.com/opensearch-project/OpenSearch/pull/7646
The below mentioned class is removed in the above PR.

```
Caused by: java.lang.NoClassDefFoundError: org/opensearch/core/common/io/stream/BaseStreamOutput
	at java.base/java.lang.Class.getDeclaredMethods0(Native Method)
	at java.base/java.lang.Class.privateGetDeclaredMethods(Class.java:3402)
	at java.base/java.lang.Class.getDeclaredMethods(Class.java:2504)
	at net.bytebuddy.description.method.MethodList$ForLoadedMethods.<init>(MethodList.java:152)
	at net.bytebuddy.description.type.TypeDescription$ForLoadedType.getDeclaredMethods(TypeDescription.java:8940)
	at net.bytebuddy.dynamic.scaffold.InstrumentedType$Factory$Default$1.represent(InstrumentedType.java:438)
	at net.bytebuddy.ByteBuddy.redefine(ByteBuddy.java:886)
	at net.bytebuddy.ByteBuddy.redefine(ByteBuddy.java:861)
	at org.mockito.internal.creation.bytebuddy.InlineBytecodeGenerator.transform(InlineBytecodeGenerator.java:390)
	at java.instrument/java.lang.instrument.ClassFileTransformer.transform(ClassFileTransformer.java:244)
	at java.instrument/sun.instrument.TransformerManager.transform(TransformerManager.java:188)
	at java.instrument/sun.instrument.InstrumentationImpl.transform(InstrumentationImpl.java:541)
	at java.instrument/sun.instrument.InstrumentationImpl.retransformClasses0(Native Method)
	at java.instrument/sun.instrument.InstrumentationImpl.retransformClasses(InstrumentationImpl.java:169)
	at org.mockito.internal.creation.bytebuddy.InlineBytecodeGenerator.triggerRetransformation(InlineBytecodeGenerator.java:281)
	... 113 more
Caused by: java.lang.ClassNotFoundException: org.opensearch.core.common.io.stream.BaseStreamOutput
	at java.base/jdk.internal.loader.BuiltinClassLoader.loadClass(BuiltinClassLoader.java:641)
	at java.base/jdk.internal.loader.ClassLoaders$AppClassLoader.loadClass(ClassLoaders.java:188)
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:520)
	... 128 more

```
